runners/direct-java/src/main/java/org/apache/beam/runners/direct/StatefulParDoEvaluatorFactory.java

```diff
@@ -244,27 +249,40 @@ public void processElement(WindowedValue<KeyedWorkItem<K, KV<K, InputT>>> gbkRes
         delegateEvaluator.processElement(windowedValue);
       }
 
+      Instant lastFired = null;
       for (TimerData timer : gbkResult.getValue().timersIterable()) {
         checkState(
             timer.getNamespace() instanceof WindowNamespace,
             "Expected Timer %s to be in a %s, but got %s",
             timer,
             WindowNamespace.class.getSimpleName(),
             timer.getNamespace().getClass().getName());
-        WindowNamespace<?> windowNamespace = (WindowNamespace) timer.getNamespace();
-        BoundedWindow timerWindow = windowNamespace.getWindow();
-        delegateEvaluator.onTimer(timer, timerWindow);
+        checkState(
+            lastFired == null || !lastFired.isAfter(timer.getTimestamp()),
+            "lastFired was %s, current %s",
+            lastFired,
+            timer.getTimestamp());
+        if (lastFired != null && lastFired.isBefore(timer.getTimestamp())) {
+          pushedBackTimers.add(timer);
+        } else {
+          lastFired = timer.getTimestamp();
+          WindowNamespace<?> windowNamespace = (WindowNamespace) timer.getNamespace();
+          BoundedWindow timerWindow = windowNamespace.getWindow();
+          delegateEvaluator.onTimer(timer, timerWindow);
```

Comment:
I'm reading this really quickly, but I think it works only for loops of the same timer, not the general case. And it also leaves the "should fire" implicit only by the fact that it already fired, instead of checking it. I think if you have a unit test where two timer callbacks set each other to "already firing" times you will uncover some issues.

I think there's a simple logic you might have:

1. Add the fired timers in the incoming bundle to a priority queue
2. While any timers to fire, call the first callback
    a. For any timer set by the `@OnTimer` callback, you check if it should fire. If yes, then add to fired timers priority queue. If no, then add it to the committed result.

Reply:
I will add the test and check if there are any issues. Regarding the logic you described - generally, yes, that would be the ideal approach. It would also solve some other issues, that still remain, even after this PR - but - it would require a larger refactoring, because there is currently no simple (or simple and apparent to me) way of knowing if `@OnTimer` method actually setup any new timer. The `delegateEvaluator` would have to signal that, and that boils down to `PushbackSideInputDoFnRunner.onTimer` would have to return the new timers. That is API change in core and I wanted to avoid that. But maybe I've overlooked  some other way.

Reply:
I have a failing test case for the case you described. Will work on that. Many thanks!

Reply:
@kennknowles Fixed the test. I'm not sure if it is exactly what you meant (although the new test really uncovered some flaws in the original implementation). Can you have a look [here](https://github.com/apache/beam/pull/9190/files#diff-37ebd0b68ae7d5b951cfca73aba52cc1R689), please?

Reply:
I don't think it is quite what I might. Here:

t1.set(50)
t2.set(60)

watermark advances to 100

t1 fires and sets t2 to 40 -- since it is moved earlier it will fire in the same bundle
t2 fires and sets t1 to 30 -- will not fire in the same bundle, but will be committed in the result?

I guess this might get the right result... I was just expecting all the timers that could fire to do it in the same bundle. It is a bit weird that the normal case of timers just coming in order and not setting anything will not fire in one bundle. Did I read that code wrong? This is why I suggested a prio queue, to remove all the logic about `lastFired` and commits. You just fire all the timers, in order (including newly set timers), until the bundle is done. I wonder if there is a use case for always setting a timer for the same time and hoping the bundle will still commit over and over again.

Reply:
Is it possible (and valid) to set timer for time that is *before* the time another timer was set for? That seems weird and actually can cause exception to be thrown in some cases.